server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java

```diff
@@ -145,7 +145,9 @@ public BulkRequest add(IndexRequest request) {
     BulkRequest internalAdd(IndexRequest request) {
         Objects.requireNonNull(request, "'request' must not be null");
         applyGlobalMandatoryParameters(request);
-
+        request.routing(valueOrDefault(request.routing(), globalRouting));
```

Comment:
It might be safer to not apply the default routing option. 

routing is exceptionally tricky and difficult to debug. Might be better to create an issue and remove it from this PR.